app/services/hts_service/workflow_engine.rb

```diff
@@ -0,0 +1,194 @@
+# frozen_string_literal: true
+
+require 'htn_workflow'
+require 'set'
+
+module HTSService
+  class WorkflowEngine
+    include ModelUtils
+
+    def initialize(patient:, date: nil, program: nil)
+      @patient = patient
+      @program = program || load_hts_program
+      @date = date || Date.today
+    end
+
+    # Retrieves the next encounter for bound patient
+    def next_encounter
+      state = INITIAL_STATE
+      loop do
+        state = next_state state
+        break if state == END_STATE
+
+        LOGGER.debug "Loading encounter type: #{state}"
+        encounter = load_states(state)
+
+        encounter_type = EncounterType.find_by(name: state)
+        encounter_type.name = encounter
+
+          return htn_transform(encounter_type) if valid_state?(state)
+      end
+
+      nil
+    end
+
+    private
+
+    LOGGER = Rails.logger
+
+    # Encounter types
+    INITIAL_STATE = 0 # Start terminal for encounters graph
+    END_STATE = 1 # End terminal for encounters graph
+    PREGNANCY_STATUS = 'PREGNANCY STATUS'
+    CIRCUMCISION = 'CIRCUMCISION'
+    SOCIAL_HISTORY = 'SOCIAL HISTORY'
+    SCREENING = 'SCREENING'
+    APPOINTMENT = 'APPOINTMENT'
+    HTS_Contact = 'HTS Contact'
+    REFERRAL = 'REFERRAL'
+    PARTNER_RECEPTION = 'Partner Reception'
+
+    # Encounters graph
+    ENCOUNTER_SM = {
+      INITIAL_STATE => PREGNANCY_STATUS,
+      PREGNANCY_STATUS => CIRCUMCISION,
+      CIRCUMCISION => SOCIAL_HISTORY,
+      SOCIAL_HISTORY => SCREENING,
+      SCREENING => APPOINTMENT,
+      APPOINTMENT => REFERRAL,
+      REFERRAL => PARTNER_RECEPTION,
+      PARTNER_RECEPTION => END_STATE
+    }.freeze
+
+    STATE_CONDITIONS = {
```

Comment:
If possible, can you have meaningful function conditions to make encounter conditions easy to identify. 

E.g

`PREGNANCY_STATUS => %I[is_female_client]`

`CIRCUMCISION => %[is_male_client]`

`SOCIAL_HISTORY => %[has_social_history]`

e.t.c